arrows/core/parameterize_mesh.cxx

```diff
@@ -0,0 +1,294 @@
+  attach_logger( "arrows.core.parameterize_mesh" );
+}
+
+// Copy constructor
+parameterize_mesh
+::parameterize_mesh(const parameterize_mesh& other)
+  : d_(new priv(*other.d_))
+{
+}
+
+
+// Destructor
+parameterize_mesh
+::~parameterize_mesh()
+{
+}
+
+
+// Get this algorithm's \link vital::config_block configuration block \endlink
+vital::config_block_sptr
+parameterize_mesh::get_configuration() const
+{
+  vital::config_block_sptr config =
+      vital::algo::parameterize_mesh::get_configuration();
+  config->set_value("resolution", d_->resolution,
+                    "Resolution used to link texture units (pixel) to mesh units (meter, ...)"
+                    "It is expressed in mesh units / pixel.");
+
+  config->set_value("interior_margin", d_->interior_margin,
+                    "It is the margin between triangles on each axis (in pixels).");
+
+  config->set_value("exterior_margin", d_->exterior_margin,
+                    "It is the exterior margin placed on each border of the texture.");
+  return config;
+}
+
+
+// Set the configuration
+void parameterize_mesh
+::set_configuration(vital::config_block_sptr in_config)
+{
+  vital::config_block_sptr config = this->get_configuration();
+
+  config->merge_config(in_config);
+
+  d_->resolution = config->get_value<double>("resolution", d_->resolution);
+  d_->interior_margin = config->get_value<unsigned int>("interior_margin", d_->interior_margin);
+  d_->exterior_margin = config->get_value<unsigned int>("exterior_margin", d_->exterior_margin);
+}
+
+
+// Check that the algorithm's configuration vital::config_block is valid
+bool parameterize_mesh
+::check_configuration(vital::config_block_sptr config) const
+{
+  return config->has_value("resolution") && config->has_value("interior_margin")
+      && config->has_value("exterior_margin");
+}
+
+
+/// Parameterize a mesh
+std::pair<unsigned int, unsigned int>
+parameterize_mesh::parameterize(kwiver::vital::mesh_sptr mesh) const
+{
+  if (mesh->faces().regularity() != 3)
+  {
+    VITAL_THROW( algorithm_exception, this->type_name(), this->impl_name(),
+                 "This algorithm expects a regular mesh with triangular faces.");
+  }
+
+  const mesh_face_array_base& faces = mesh->faces();
+  const mesh_vertex_array<3>& vertices = dynamic_cast<const mesh_vertex_array<3>&>(mesh->vertices());
+
+  // Map each triangle in 2d. The longest edge should be horizontal and its left point is (0, 0)
+  std::vector<triangle_t> triangles(mesh->num_faces());
+  double total_area = 0.0;
+  for (unsigned int f = 0; f < mesh->num_faces(); ++f)
+  {
+    // face 3d points
+    vector_3d pt1 = vertices[faces(f, 0)];
+    vector_3d pt2 = vertices[faces(f, 1)];
+    vector_3d pt3 = vertices[faces(f, 2)];
+
+    // triangle edges
+    vector_3d pt1pt2 = pt2 - pt1;
+    vector_3d pt1pt3 = pt3 - pt1;
+    vector_3d pt2pt3 = pt3 - pt2;
+
+    if (pt1pt2.norm() < 1e-5 || pt1pt3.norm() < 1e-5 || pt2pt3.norm() < 1e-5 || (pt1pt2.cross(pt1pt3)).norm() < 1e-5)
```

Comment:
This check was done to ignore very thin triangles which would not appear in the atlas because they are smaller than a pixel at the given resolution. But, yes, to be completely invariant to the scale of the mesh we should remove this.